addons/account/models/account_move_line.py

```diff
@@ -1121,6 +1139,33 @@ def _compute_analytic_distribution(self):
                 })
                 line.analytic_distribution = distribution or line.analytic_distribution
 
+    @api.depends('discount_date', 'date_maturity')
+    def _compute_payment_date(self):
+        for line in self:
+            line.payment_date = line.discount_date if line.discount_date and date.today() <= line.discount_date else line.date_maturity
+
+    def _search_payment_date(self, operator, value):
+        assert operator == '='
+        return ['|', '&', ('discount_date', '>=', str(date.today())), ('discount_date', '<=', value), ('date_maturity', '<=', value)]
```

Comment:
```suggestion
        return [
            '|', 
            '&', ('discount_date', '>=', str(date.today())), ('discount_date', '<=', value), 
            '&', ('discount_date', '<', str(date.today())), ('date_maturity', '<=', value),
        ]
```

We need to search on either one either the other field, so the clauses on today need to be opposite to each other.
Also need to check if it is working as expected when `discount_date` is not set?

Also, add a test case for this (create multiple lines with various dates), and do multiple searches with mocked today (use `freeze_time`)